### ISSUE TITLE
Unit tests not working

### ISSUE THREAD

**Issue by codeanticode** [rawnsley/HubsCloudUtil#1]
@rawnsley This library could be very useful for programmatic access of hubs spaces and assets, thanks for making this code available.

I can build the library using gradle, however, the five unit tests are giving errors. I post below the exception and standard error log from the RoomConnectionTest.joinAndLeave test, which does not require authentication:

```
java.lang.Exception: Failed to connect to reticulum server https://hubs.mozilla.com
	at avn.portal.ReticulumConnection.<init>(ReticulumConnection.kt:88)
	at avn.portal.ReticulumConnectionTest.openAndClose(ReticulumConnectionTest.kt:16)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

```
[INFO] ReticulumConnection https://hubs.mozilla.com connecting...
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com metadata: version: 1.0.20220322172821; pool: arbre; phx_host: nervous-werewolf.reticulum.io' phx_port: 4000
[INFO] ReticulumConnection https://hubs.mozilla.com connecting at endpoint wss://nervous-werewolf.reticulum.io/socket/websocket
[INFO] ReticulumConnection https://hubs.mozilla.com waiting for phoenix socket connection...
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
```

The room in this test, [w6QJKv9](https://hubs.mozilla.com/w6QJKv9), is valid, so I'm wondering what I might be missing. In all other tests, the key issue seems to be the same, "Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure". Thanks again!

**Comment by rawnsley** [comment]
Thanks for flagging this. I can reproduce the problem on MacOS and it is failing to create the WebSocket to the Hubs reticulum server. The same socket connection in wscat or Postman works fine. It has definitely worked in the past so it's a bit of a mystery.

**Comment by rawnsley** [comment]
Even a [simple WebSocket connect](https://github.com/rawnsley/HubsCloudUtil/commit/4f16d5fc73248e2c665a633e5b2280c12863fa56) is failing to handshake so it's not specific to the Hubs or Phoenix code. It might be something to do with the OkHttp client configuration.

Enabling [more network debugging](https://github.com/rawnsley/HubsCloudUtil/commit/25e2d263a25f4fa1515358d9acab999a8457df94#diff-7e478e96fb570ce2e15f5928fcd43f9f78f83138f0eb526258a33d00a8953e6cR33) shows me the ClientHello and server response failure in more detail, but there is no obvious reason given.

Almost identical HTTPS negotiations with hubs.mozilla.com using the same HTTP client work fine. Similarly, connecting to wss://echo.websocket.events works fine with the same code.

There is something specific about the reticulum server that is preventing the connection, but I can't see it.

Some useful information I've found while investigating:

https://stackoverflow.com/a/6353956/671393 
https://www.ssllabs.com/ssltest/analyze.html?d=peaceful-hobbit.reticulum.io

**Comment by codeanticode** [comment]
Thanks for looking into it so quickly! It must have been caused by change in the reticulum code between now and the last time your library worked correctly... or maybe a change in the AWS configuration? The SSL report you linked at the end does not seem to reveal anything suspicious, right? I guess it's hard to test unless you deploy your own reticulum instance.

**Comment by rawnsley** [comment]
I think you are right; it seems unlikely that the Java HTTP code has changed so perhaps a change in the AWS configuration has exposed an incompatibility that somehow doesn't affect wscat or Postman. The same code appears to be working on Android so I'm struggling to imagine where it might be coming from.

What platform are you running on?

**Comment by codeanticode** [comment]
I'm on Mac right now, but could take a look on Windows and Linux later

**Comment by codeanticode** [comment]
Trying to build the library both on Linux and Windows gives this error:

```
General error during conversion: Unsupported class file major version 61
```

I have OpenJDK 17 on Linux (Ubuntu) and tried both JDK 1.8 and 17 on Windows.

**Comment by rawnsley** [comment]
@codeanticode Thanks for trying on this. I don't have a Windows machine to build on at the moment, but I'll try it next time I can.

**Comment by codeanticode** [comment]
Ok, there is no rush on my end, but will give it another try to see if I can build and run the lib on Windows and/or Linux. Maybe I need to install JDK 11? But used JDK 17 to build on Mac, so not sure why it did not work on Windows and Linux. Thank you!

**Comment by codeanticode** [comment]
@rawnsley The solution (on Linux at least) was to update the gradle distribution used by gradlew to 7.3.1. Then I was able to build and run the tests, which failed as before:

```
avn.portal.HubsApiTest > sceneChange1 FAILED
    java.lang.Exception at HubsApiTest.kt:30

avn.portal.HubsApiTest > sceneChange2 FAILED
    java.lang.Exception at HubsApiTest.kt:35

avn.portal.HubsApiTest > sendMessageSimple FAILED
    java.lang.Exception at HubsApiTest.kt:25

avn.portal.ReticulumConnectionTest > openAndClose FAILED
    java.lang.Exception at ReticulumConnectionTest.kt:16

avn.portal.RoomConnectionTest > joinAndLeave FAILED
    java.lang.Exception at RoomConnectionTest.kt:19

6 tests completed, 5 failed
```
Actually it says 6 tests completed, not sure which one is the 6th one... Anyways here you have the detailed error mesage from JoinAndLeave:

```
java.lang.Exception: Failed to connect to reticulum server https://hubs.mozilla.com
	at avn.portal.ReticulumConnection.<init>(ReticulumConnection.kt:88)
	at avn.portal.RoomConnectionTest.joinAndLeave(RoomConnectionTest.kt:19)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

and the standard error output:

```
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com connecting...
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com metadata: version: 1.0.20220328174507; pool: earth; phx_host: epic-orc.reticulum.io' phx_port: 4000
[INFO] ReticulumConnection https://hubs.mozilla.com connecting at endpoint wss://epic-orc.reticulum.io/socket/websocket
[INFO] ReticulumConnection https://hubs.mozilla.com waiting for phoenix socket connection...
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Socket attempting to reconnect
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
[INFO] ReticulumConnection https://hubs.mozilla.com socket closed
[WARNING] ReticulumConnection https://hubs.mozilla.com unexpected null phoenixSocket
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: error javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
[ERROR] ReticulumConnection https://hubs.mozilla.com socket error: Received fatal alert: handshake_failure null
[INFO] ReticulumConnection https://hubs.mozilla.com socket: Transport: close
```
Seems like it's the same issue as in macOS. I will see if I can test on Windows as well.

**Comment by codeanticode** [comment]
Ok, same error in the unit tests om Windows after upgrading to gradle 7.3.1. Hope this is helpful to debug the issue, and happy to run some other tests if needed. Also, created a PR https://github.com/rawnsley/HubsCloudUtil/pull/2 with the gradle update.

**Comment by rawnsley** [comment]
Thanks for looking into this, and thanks for the PR. I'll have another swing at it, but it's quite a puzzle.

**Comment by codeanticode** [comment]
Cool, thanks. I wonder if testing with a Hub Clouds deployment where one can tweak the parameters could be helpful to determine what's causing the issue.

**Comment by rawnsley** [comment]
I think this change https://github.com/rawnsley/HubsCloudUtil/commit/cb27ef07e08e57def63b2b6bfabfc68af9492ac8 fixes the problem, at least on my Mac. I need to test it more widely.

I was driven back here because I updated an AWS AMI from Java 1.8.0_265 to 1.8.0_312 and the same problem started happening. I stumbled upon a few interesting issues during the investigation:

- Similar sounding bug https://bugs.openjdk.java.net/browse/JDK-8241360
- Explicit Erlang bug http://erlang.org/pipermail/erlang-questions/2020-March/099201.html

I suspect it's a quirk of the interaction between Erlang and this later build of Java 1.8. I wasn't able to narrow down exactly what was going wrong, partly because I know nothing about Erlang. I did however find that swapping OpenJDK for Conscrypt appears to fix the problem.
